scripts/decompose_flow_vectors

```diff
@@ -14,38 +14,102 @@ import sys
+class InvalidVariableError(Exception):
+    pass
+
+
+class VariableNotFoundError(InvalidVariableError):
+    pass
+
+
+class NotAFlowVariableError(InvalidVariableError):
+    pass
+
+
+class NotAVicModelError(InvalidVariableError):
+    pass
 
-def main(args):
-    # check that source file is usable:
-    source = Dataset(args.source_file, "r", format="NETCDF4")
+
+def source_check(source):
+    source_file_path = source.filepath()
 
     if not "lon" in source.dimensions or not "lat" in source.dimensions:
-        logger.debug(
+        logger.critical(
             "{} does not have latitude and longitude dimensions".format(
-                args.source_file
+                source_file_path
             )
         )
+        raise NoDimensionsError
+
+    valid_variables = []
+    for v in source.variables:
+        variable = source.variables[v]
+        if (
+            hasattr(variable, "dimensions")
+            and "lon" in variable.dimensions
+            and "lat" in variable.dimensions
+        ):
+            if np.ma.max(variable[:]) <= 9 and np.ma.min(variable[:]) >= 1:
+                valid_variables.append(v)
+
+    if len(valid_variables) == 0:
+        logger.critical(
+            "{} does not have a valid flow variable".format(source_file_path)
+        )
         source.close()
-        sys.exit()
+        raise NoFlowVariablesError
 
-    if not args.variable in source.variables:
-        logger.debug(
-            "Variable {} is not found in {}".format(args.variable, args.source_file)
+
+def variable_check(source, variable):
+    source_file_path = source.filepath()
+
+    if not variable in source.variables:
+        logger.critical(
+            "Variable {} is not found in {}".format(variable, source_file_path)
         )
         source.close()
-        sys.exit()
+        raise VariableNotFoundError
 
-    flow_variable = source.variables[args.variable]
+    flow_variable = source.variables[variable]
 
     if not "lon" in flow_variable.dimensions or not "lat" in flow_variable.dimensions:
-        logger.debug("Variable {} is not associated with a grid".format(args.variable))
+        logger.critical("Variable {} is not associated with a grid".format(variable))
         source.close()
-        sys.exit()
+        raise NotAFlowVariableError
 
     if np.ma.max(flow_variable[:]) > 9 or np.ma.min(flow_variable[:]) < 1:
-        logger.debug("Variable {} is not a valid flow routing".format(args.variable))
+        logger.critical("Variable {} is not a valid flow routing".format(variable))
         source.close()
-        sys.exit()
+        raise NotAVicModelError
+
+
+def main(args):
+    # check that source file is usable:
+    source = Dataset(args.source_file, "r", format="NETCDF4")
+
+    try:
+        source_check(source)
+    except InvalidSourceError:
+        sys.exit(1)
+
+    try:
+        variable_check(source, args.variable)
+    except InvalidVariableError:
+        sys.exit(2)
```

Comment:
One one hand, I can see the merit to being as specific as possible (via the exit code) as to why the program exited. However, at this point (correct me if I'm wrong), I don't think that we're automating the use of this script from bash and there might not be any use for these specific exit code.

Which then raises the question as to whether there's a purpose for creating a class hierarchy above for the Exception types. Ever since attending this [provocatively named, and humorously informative PyCon talk](https://www.youtube.com/watch?v=o9pEzgHorH0) "Stop Writing Classes", I have been hyper-skeptical of custom exception types. Especially since there are a large number of built-in Python `Exception`s and that most `Exception` children allow instantiation with an informative message.

So I'd advise that you probably don't need almost 30 lines of custom Exception hierarchy and can probably do raising one of the [builtin Exceptions](https://docs.python.org/3/library/exceptions.html) (`TypeError`, `ValueError`, `RuntimeError`, ...) with an informative message.

Reply:
This strategy has been selected based on my plan for applying the functions in [wps_decompose_flow_vectors](https://github.com/pacificclimate/thunderbird/blob/issue/12/thunderbird/processes/wps_decompose_flow_vectors.py#L123). I wanted the log message to appear on WPS environment as well, therefore the distinctions between error conditions were needed. However, instead of creating two main user-defined exceptions, it would be wiser to make the lower level exceptions to inherit built-in exceptions like you suggested.

Reply:
Yeah, having the log message appear in the WPS environment is absolutely an excellent use case. But I think that's what the `message` attribute of an `Exception` is for.

Reply:
I would agree with James' comments here.